src/qibolab/tests/test_instruments_qmsim.py

```diff
@@ -110,18 +112,56 @@ def test_qmsim_qubit_spectroscopy(simulator):
     assert_regression(samples, "qubit_spectroscopy")
 
 
-@pytest.mark.parametrize("qubits", [[1, 2], [2, 3]])
-def test_qmsim_bell_circuit(simulator, qubits):
-    backend = QibolabBackend(simulator)
-    circuit = Circuit(5)
-    circuit.add(gates.H(qubits[0]))
-    circuit.add(gates.CNOT(*qubits))
-    circuit.add(gates.M(*qubits))
-    result = backend.execute_circuit(circuit, nshots=1, check_transpiled=True)
-    result = result.execution_result
+gatelist = [
+    ["I", "I"],
+    ["RX(pi)", "RX(pi)"],
+    ["RY(pi)", "RY(pi)"],
+    ["RX(pi)", "RY(pi)"],
+    ["RY(pi)", "RX(pi)"],
+    ["RX(pi/2)", "I"],
+    ["RY(pi/2)", "I"],
+    ["RX(pi/2)", "RY(pi/2)"],
+    ["RY(pi/2)", "RX(pi/2)"],
+    ["RX(pi/2)", "RY(pi)"],
+    ["RY(pi/2)", "RX(pi)"],
+    ["RX(pi)", "RY(pi/2)"],
+    ["RY(pi)", "RX(pi/2)"],
+    ["RX(pi/2)", "RX(pi)"],
+    ["RX(pi)", "RX(pi/2)"],
+    ["RY(pi/2)", "RY(pi)"],
+    ["RY(pi)", "RY(pi/2)"],
+    ["RX(pi)", "I"],
+    ["RY(pi)", "I"],
+    ["RX(pi/2)", "RX(pi/2)"],
+    ["RY(pi/2)", "RY(pi/2)"],
+]
+
+
+@pytest.mark.parametrize("count,gate_pair", enumerate(gatelist))
+def test_qmsim_allxy(simulator, count, gate_pair):
+    qubits = [1, 2, 3, 4]
+    sequence = PulseSequence()
+    for qubit in qubits:
+        start = 0
+        for gate in gate_pair:
+            if gate == "I":
+                pulse = None
+            elif gate == "RX(pi)":
+                pulse = simulator.create_RX_pulse(qubit, start=start)
+            elif gate == "RX(pi/2)":
+                pulse = simulator.create_RX90_pulse(qubit, start=start)
+            elif gate == "RY(pi)":
+                pulse = simulator.create_RX_pulse(qubit, start=start, relative_phase=np.pi / 2)
+            elif gate == "RY(pi/2)":
+                pulse = simulator.create_RX90_pulse(qubit, start=start, relative_phase=np.pi / 2)
```

Comment:
This is subjective and disputable, but I would find it clearer written as:
```py
# top-level
PULSES = {
  "I": lambda **kwargs: None,
  "RX(pi)": lambda qubit, *, start, **kwargs: simulator.create_RX90_pulse(qubit, start=start)
  ...
}
```
and then:
```suggestion
            pulse = PULSES[gate](qubit, start=start, relative_phase=np.pi / 2)
```

Reply:
I agree, I will fix this and the comments above when I implement the changes for hdf5. Just noting that this method, along with some other tests implemented in this PR, was copied [from qibocal](https://github.com/qiboteam/qibocal/blob/9a86c0aeb26167ad60f1177aa11e31fc74319c5f/src/qibocal/calibrations/characterization/allXY.py#L363). At some point we may need to think of ways to reduce this code duplication but does not seem very trivial to me right now.

Reply:
> The practical solution I'd propose for the current approach is to reduce the duplication as much as possible: whenever you have something duplicated, try to decide whether it belongs to one or the other. In case you have multiple non-trivial examples, we can start explicitly discussing them.

For this case, the code belongs to qibocal. What I am trying to do here is test if some (ideally we would like to test all) of qibocal routines and also some qibo circuits, are sending the pulse waveforms they are supposed to, before I deploy them to real qubits.

This is useful because if something goes wrong in characterization, we can use it to check if it's due to a bug in the driver (which is among the most usual reasons something goes wrong). I guess we have not been doing it so far because qblox did not provide such simulator so it is not trivial to check the output waveforms, but QM does. 

Maybe these tests belong to qibocal, however they are not hardware agnostic, so I am not sure.

Reply:
So, if I correctly see the arrows, then qibolab is a qibocal user. In this case, we can have:

- a qibocal module to manage simulators, including `PULSES` dict and related utilities, part of the public API of qibocal
- tests in qibocal to use the content of the module
- tests in qibolab to use the same content

Of course this means that we need to use qibocal inside qibolab also for testing, so we should agree on some interface and keep it stable.
For the time being the duplication is fine, then we will remove in this way at a later stage.

If you agree on the strategy, can you open a corresponding issue on qibocal? @stavros11

If instead this is not the case, they should go inside qibolab. Out of the two pacakges, one should not depend on the other, so there has to be a clear arrow: either qibocal is a qibolab user, and the platform is fully calibration independent, or the other way round. The only further option is to have both independent from each other, and use together in a third package (but this I would avoid, definitely now and possibly even later on).

Reply:
Regarding dependencies, I think the way it is now is the right one: qibolab is a dependency of qibocal.
* qibolab defines various objects (`Pulse`, `PulseSequence`, `Platform`, `Qubit` and soon `Sweeper`) that allow executing arbitrary sequences of pulses using different instrument setups.
* qibocal uses these objects to execute specific pulse sequences that are commonly used to characterize and calibrate superconducting qubits and save and post-process the acquired results.

Under these descriptions the arrow should be clear and I think that reverting would require major design changes on both sides.

Here I am cheating because I am using the pulse sequences that qibocal (the "client") wants to execute to test my driver. I could have used other random pulse sequences (and maybe I should). I guess the tests I have here will be deprecated once we find a way to test all calibration routines on different platforms directly on qibocal.

Reply:
> Here I am cheating because I am using the pulse sequences that qibocal (the "client") wants to execute to test my driver. I could have used other random pulse sequences (and maybe I should). I guess the tests I have here will be deprecated once we find a way to test all calibration routines on different platforms directly on qibocal.

Then I agree: if qibocal is the client, either we test it there with its routines, or we move some code here, and use there from qibolab.
But, given the arrow direction, I'd say that qibolab should test drivers at a rather shallow level, while more involved tests will be delegated to qibocal.
In a sense, since it is missing the *advanced tools*, then its task consist in just providing a barely working platform. How well it is working is something that will be assessed in qibocal.